src/transformStateWithClones.js

```diff
@@ -7,7 +7,31 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const result = [];
+  let currentState = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'clear':
+        currentState = {};
+        break;
+      case 'addProperties':
+        currentState = { ...currentState, ...action.extraData };
+        break;
+      case 'removeProperties':
+        currentState = { ...currentState };
+
+        for (const key of action.keysToRemove) {
+          delete currentState[key];
+        }
+        break;
+      default:
+        throw new Error(`Unknown action type: ${action.type}`);
+    }
+    result.push({ ...currentState });
```

Comment:
The copying of `currentState` into the `result` array is done correctly to avoid mutation of the objects within the result array. However, this operation could be performed once after the switch statement to adhere to the DRY (Don't Repeat Yourself) principle.